src/zc/buildout/easy_install.py

```diff
@@ -518,98 +570,76 @@ def _get_dist(self, requirement, ws, for_buildout_run=False):
+            self._env_rescan_dest()
             dist = self._env.best_match(requirement, ws)
             logger.info("Got %s.", dist)
 
         else:
             dists = [dist]
 
         if not self._install_from_cache and self._use_dependency_links:
-            for dist in dists:
-                if dist.has_metadata('dependency_links.txt'):
-                    for link in dist.get_metadata_lines('dependency_links.txt'):
-                        link = link.strip()
-                        if link not in self._links:
-                            logger.debug('Adding find link %r from %s',
-                                         link, dist)
-                            self._links.append(link)
-                            self._index = _get_index(self._index_url,
-                                                     self._links,
-                                                     self._allow_hosts)
+            self._add_dependency_links_from_dists(dists)
 
         if self._check_picked:
-            # Check whether we picked a version and, if we did, report it:
-            for dist in dists:
-                if not (
-                    dist.precedence == pkg_resources.DEVELOP_DIST
-                    or
-                    (len(requirement.specs) == 1
-                     and
-                     requirement.specs[0][0] == '==')
-                    ):
-                    logger.debug('Picked: %s = %s',
-                                 dist.project_name, dist.version)
-                    self._picked_versions[dist.project_name] = dist.version
-
-                    if not self._allow_picked_versions:
-                        raise zc.buildout.UserError(
-                            'Picked: %s = %s' % (dist.project_name,
-                                                 dist.version)
-                            )
+            self._check_picked_requirement_versions(requirement, dists)
 
         return dists
 
-    def _maybe_add_setuptools(self, ws, dist):
-        if dist.has_metadata('namespace_packages.txt'):
-            for r in dist.requires():
-                if r.project_name in ('setuptools', 'setuptools'):
-                    break
-            else:
-                # We have a namespace package but no requirement for setuptools
-                if dist.precedence == pkg_resources.DEVELOP_DIST:
-                    logger.warn(
-                        "Develop distribution: %s\n"
-                        "uses namespace packages but the distribution "
-                        "does not require setuptools.",
-                        dist)
-                requirement = self._constrain(
-                    pkg_resources.Requirement.parse('setuptools')
-                    )
-                if ws.find(requirement) is None:
-                    for dist in self._get_dist(requirement, ws):
-                        ws.add(dist)
+    def _add_dependency_links_from_dists(self, dists):
```

Comment:
> ... I wouldn't have split these out because the condition introducing the code made the intent as clear as a function name.
> 
> But the calling function and these aren't very small, so :)

Yes, it's one of those multi-page functions. Extracting these methods made the function almost fit in one screen in my editor.